### PR TITLE
Integrate duplicated PoW testing code into a file

### PR DESCRIPTION
**Pull request by marktwtn**
The selected hardware acceleration are added by user-defined macro.

Output message:
```
$ make check:
*** Validating build/test-trinary ***
        [ Verified ]
*** Validating build/test-curl ***
        [ Verified ]
*** Validating build/test-dcurl ***
        [ Verified ]
*** Validating build/test-multi-pow ***
        [ Verified ]
*** Validating build/test-pow ***
CPU - SSE
Success.
```

### CODE REVIEW

File: tests/test-pow.c [hunk 1]
```diff
@@ -49,31 +81,58 @@ int main()
...
-    freeTrobject(hash_trytes);
-    freeTrobject(ret_trits);
+
+    for (int idx = 0; idx < sizeof(ImplContextArr)/sizeof(ImplContext); idx++) {
+        printf("%s\n",description[idx]);
+
+        ImplContext *PoW_Context_ptr = &ImplContextArr[idx];
+
+        /* test implementation with mwm = 14 */
+        initializeImplContext(PoW_Context_ptr);
```
Comment (furuame) [inline tests/test-pow.c:109]: The ImplContext should also be freed by `destroyImplContext()`

File: tests/test-pow.c [hunk 2]
```diff
@@ -49,31 +81,58 @@ int main()
...
+        void *pow_ctx = getPoWContext(PoW_Context_ptr, (int8_t *) trytes, mwm);
+        assert(pow_ctx);
+        doThePoW(PoW_Context_ptr, pow_ctx);
+        int8_t *ret_trytes = getPoWResult(PoW_Context_ptr, pow_ctx);
+        assert(ret_trytes);
+        freePoWContext(PoW_Context_ptr, pow_ctx);
+
+        Trytes_t *trytes_t = initTrytes(ret_trytes, 2673);
```
Comment (furuame) [inline tests/test-pow.c]: Replace the constant `2673` with marco defined in `constants.h`

File: tests/test-pow.c [hunk 3]
```diff
@@ -49,31 +81,58 @@ int main()
...
+        assert(trytes_t);
+        Trytes_t *hash_trytes = hashTrytes(trytes_t);
+        assert(hash_trytes);
+        Trits_t *ret_trits = trits_from_trytes(hash_trytes);
+        assert(ret_trits);
+
+        /* Validation */
+        for (int i = 243 - 1; i >= 243 - mwm; i--) {
```
Comment (furuame) [inline tests/test-pow.c]: `243` should also be replaced.

File: tests/test-pow.c [hunk 4]
```diff
@@ -1,8 +1,40 @@
...
+
+#if defined(ENABLE_OPENCL)
+extern ImplContext PoWCL_Context;
+#endif
+
+#if defined(ENABLE_FPGA_ACCEL)
 extern ImplContext PoWFPGAAccel_Context;
+#endif
+
+const char *description[ ] = {
```
Comment (furuame) [inline tests/test-pow.c:21]: Please fix the coding style like the following:

```
const char *description[] = {
#if defined(ENABLE_AVX)
    "CPU - AVX",
#elif defined(ENABLE_SSE)
    ........
}

File: tests/test-pow.c [hunk 5]
```diff
@@ -49,31 +81,58 @@ int main()
...
-    assert(trytes_t);
-    Trytes_t *hash_trytes = hashTrytes(trytes_t);
-    assert(hash_trytes);
-
-    /* Validation */
-    Trits_t *ret_trits = trits_from_trytes(hash_trytes);
-    for (int i = 243 - 1; i >= 243 - mwm; i--) {
-        assert(ret_trits->data[i] == 0);
-    }
+    ImplContext ImplContextArr[ ] = {
```
Comment (furuame) [inline tests/test-pow.c:84]: ditto.

File: tests/test-pow.c [hunk 6]
```diff
@@ -49,31 +81,58 @@ int main()
...
+        PoWFPGAAccel_Context,
+        #endif
+    };
 
-    free(ret_trytes);
-    freeTrobject(trytes_t);
-    freeTrobject(hash_trytes);
-    freeTrobject(ret_trits);
+
+    for (int idx = 0; idx < sizeof(ImplContextArr)/sizeof(ImplContext); idx++) {
```
Comment (furuame) [inline tests/test-pow.c]:
```
for (int idx = 0; idx < sizeof(ImplContextArr) / sizeof(ImplContext); idx++) {
```

**Review by furuame — CHANGES_REQUESTED** [review]

**Comment by furuame** [comment]
Could you attach the output message (after running `make check` in your new commit) in your PR's description?

File: tests/test-pow.c [hunk 7]
```diff
@@ -49,31 +81,58 @@ int main()
...
-    Trytes_t *hash_trytes = hashTrytes(trytes_t);
-    assert(hash_trytes);
-
-    /* Validation */
-    Trits_t *ret_trits = trits_from_trytes(hash_trytes);
-    for (int i = 243 - 1; i >= 243 - mwm; i--) {
-        assert(ret_trits->data[i] == 0);
-    }
+    ImplContext ImplContextArr[ ] = {
+        #if defined(ENABLE_AVX)
```
Comment (jserv) [inline tests/test-pow.c]: Don't indent `#if` and `#endif` statements. Always leave them without indentation (no leading space).

File: tests/test-pow.c (hunk 5 above)
Comment (jserv) [inline tests/test-pow.c:84]: In this way, you have to enumerate two sets of implementations. Can you reduce the amount of `#if` and `#endif` statements by some [C99 tricks](https://blog.noctua-software.com/c-tricks.html)?

**Comment by jserv** [comment]
This pull request was meant to resolve issue #60. You shall append `Close #60` in the end of Git commit message.

File: tests/test-pow.c (hunk 2 above)
Comment (marktwtn) [reply tests/test-pow.c]: No macro is equal to `2673` in `constants.h`.
However, there exists a macro `#define TRANSACTION_LENGTH 2673 * 3`.

Should I change the macro like this
```
#define TRANSACTION_TRYTES_LENGTH 2673
...
#define TRANSACTION_TRITS_LENGTH TRANSACTION_TRYTES_LENGTH * 3
```
or is there any suggestion?

Comment (jserv) [reply tests/test-pow.c]: I agree with the proposed changes by @marktwtn. Also, we can follow file `trit_tryte.h` in [entangled](https://github.com/iotaledger/entangled).

**Pull request merged**